### DISCUSSION TITLE
Auto center on screen code

### DISCUSSION THREAD

**Discussion by rraafays** [shortcuts/no-neck-pain.nvim#458] [Show and tell]
## Describe the problem

I tend to code on narrow displays (4:3, 5:4) out of preference but for my on the go setup I use your plugin to auto center code using a function I wrote. At the moment the function I wrote only works on mac os due to the way `wc` outputs on osx.  

## Describe the solution

I would love it if you could natively include this in a more proper way rather than my 5 minute scripting job, I'm not very good with lua but here is the function.

```
vim.api.nvim_create_user_command('Center',function()
    local command = io.popen("wc -L " .. vim.fn.expand("%:p") .. " | tr -s ' ' | cut -d ' ' -f 2")
    local width = command:read("*a")
    print(width)
    vim.cmd("NoNeckPainResize " .. width + 8)
end,{})
```

Essentially all I am doing is getting the length of the longest line of the current buffer via `wc` cutting out any irellevant output other than that value via `cut` and then resizing via your plugin + some extra width to be certain and account for the numberline (another thing I think your plugin should take into account by default).

Thanks :-)

**Comment by shortcuts** [comment]
Hey, thanks for the suggestion and using the plugin :D 

So if I summarize, what you'd like is that `NoNeckPain` resizes to the width of the longest line in the buffer when entering in it?

Do you also expect a resize if the line gets deleted/shortened/extended?

---

This seems pretty specific but I can indeed understand the motivation especially if you rarely change buffers etc.

**Comment by rraafays** [comment]
Hey, no problem I love your plugin.

Yeah so either provide a command to resize to longest line (ill just call it center) or even better if you could create an automatic center command that would be incredible I hadn't even thought of that, I actually use my center function then I have '-' & '=' bound to your width up & down

**Comment by rraafays** [comment]
I would love the automatic resizing you suggested if that's possible that's pretty much a perfect solution but even centering on BufEnter would be cool with me :-)

**Comment by rraafays** [comment]
another thing i might add is that the arbitrary extra 8 width i'm adding to the result of the `wc` command is due to this plugin including the margin in the total width. I don't know if this is doable but I think this plugin would be perfect outside of my afformentioned weird usecase if the width accounted for the margin meaning that a width of 80 means 80 characters after the margin rather than it being 80 characters wide total including the margin if that makes sense.

Thanks,
:-)

**Comment by rraafays** [comment]
update: hey, i kind of found my own solution by modifying my script, i have also made it cross platform by using the ```wc``` command via https://github.com/uutils/coreutils crate.

```
DOCUMENT_WIDTH = 80
RECEIPT_WIDTH = 60

vim.api.nvim_create_user_command("Center", function()
    local filetype = vim.bo.filetype
    if filetype == "xxd" or filetype == "alpha" then
        vim.cmd("NoNeckPainResize " .. DOCUMENT_WIDTH)
        return
    elseif filetype == "text" then
        vim.cmd("NoNeckPainResize " .. RECEIPT_WIDTH)
        return
    end
    local path = vim.fn.expand("%:p")
    local command = io.popen("coreutils wc -L " .. path:gsub(" ", "\\ "))
    if command ~= nil then
        local result = command:read("*a")
        for i in string.gmatch(result, "%S+") do
            if tonumber(i) ~= nil then vim.cmd("NoNeckPainResize " .. i + 2) end
        end
    end
end, {})

vim.api.nvim_create_autocmd("BufWritePost", {
    pattern = "",
    command = "Center",
})
```

the most notable changes are the uses of either document width or receipt width as arbitrary values to fall back on as well as autocmd to center against a newly written buffer's file.

this implementation also works with file paths that include spaces and it also assumes that no gutter, numberline or symbols line is being used as per my new config.
:-)

**Comment by rraafays** [comment]
beautiful, centered, painless
<img width="1440" alt="Screenshot 2024-01-03 at 19 02 48" src="https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/bced888b-ae7e-497b-948a-756d5b6162e0">
<img width="1440" alt="Screenshot 2024-01-03 at 19 03 00" src="https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/06aad05b-7f20-46ae-b066-f72da7daeb41">
<img width="1440" alt="Screenshot 2024-01-03 at 19 04 00" src="https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/b5d00d83-e550-4dfa-b2a9-7fb7ceabeb30">
<img width="1440" alt="Screenshot 2024-01-03 at 19 04 20" src="https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/f051a867-9be2-4035-ae84-75e6028f9b2a">

**Comment by shortcuts** [comment]
Hey, that's awesome and looks super clean! I'll definitely look for a native implementation in Neovim in order to provide this as an option, thanks!!!

**Comment by chelsea6502** [comment]
This auto-centering is super cool! What would make it even better is to set a maximum width, in case there are any rare super-long lines.

**Comment by rraafays** [comment]
huge update! figured out how to auto center
i have found a way to repurpose my original center function to save a temporary file and center against that while editing, solving the original aim of this issue.

it works by saving a temporary file in the users .local dir (currently assumes youre using lunarvim) and then runs wc on that file and passing the result to NoNeckPainResize with an arbitrary addition of 2 which seems to make the centering perfect.

there are 4 autocmds used to achieve this, both CursorMoved and CursorMovedI to constantly center while editing as well as after writing with BufWritePost since formatters might change the max width of the code.
finally, the Filetype autocmd is used to make sure that man pages are kept at a comfortable square width.     

```
            vim.api.nvim_create_user_command("Center", function()
                local filetype = vim.bo.filetype
                if filetype == "TelescopePrompt" or filetype == "alpha" or filetype == "lazy" then
                    return
                end
                vim.cmd("w! ~/.cache/lvim/width")
                local command = io.popen("wc -L " .. "~/.cache/lvim/width")
                if command ~= nil then
                    local result = command:read("*a")
                    for i in string.gmatch(result, "%S+") do
                        if tonumber(i) ~= nil then vim.cmd("NoNeckPainResize " .. i + 2) end
                    end
                end
            end, {})

            vim.api.nvim_create_user_command("Square", function()
                local command = io.popen("tput lines")
                if command ~= nil then
                    local result = command:read("*a")
                    vim.cmd("NoNeckPainResize " .. result * 2)
                end
            end, {})

            vim.api.nvim_create_autocmd("CursorMovedI", {
                pattern = "",
                command = "Center",
            })

            vim.api.nvim_create_autocmd("TextChanged", {
                pattern = "",
                command = "Center",
            })

            vim.api.nvim_create_autocmd("BufWritePost", {
                pattern = "",
                command = "Center",
            })

            vim.api.nvim_create_autocmd("FileType", {
                pattern = "man",
                command = "Square",
            })
        end
```

**Comment by rraafays** [comment]
https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/27eeab29-285c-4ebd-8694-ec79d38b5446

**Comment by rraafays** [comment]
https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/89c25fec-e9eb-4c57-b4fb-0f6358d7ff62

**Comment by rraafays** [comment]
https://github.com/shortcuts/no-neck-pain.nvim/assets/75220053/6264876a-4cbc-4a09-babb-af5286d2f023

**Comment by shortcuts** [comment]
This is awesome @rraafays!!!!!!! I love the idea and the outcome, I've started some work on the next branch in order to make nnp more responsive and extensible, if this can't be done natively with nvim until then, I'll make sure to ease your autocmds registering!

**Comment by rraafays** [comment]
Hey how's it going?
I've recently finished my industry year and I'll be free for about a month before I start University again.

I've not stopped using your plugin at all doubly so because of my auto center script, I was wondering if you got any progress on this? I'm not the best at lua but I'd love to contribute if I can! 

(this plugin with my auto center shenanigans has been so insanely robust, eye pleasing and most importantly pain free. I even get compliments about how clean my auto centering editor looks when I share my screen at work lol)

**Comment by shortcuts** [comment]
> Hey how's it going?

Wonderful and you?

> I've recently finished my industry year and I'll be free for about a month before I start University again.
> 
> 
> 
> I've not stopped using your plugin at all doubly so because of my auto center script, I was wondering if you got any progress on this? I'm not the best at lua but I'd love to contribute if I can!

Congrats and thanks for keeping up with the plugin!! That's cool to hear

If you want to tackle this feature addition, please go ahead! I've planned to add it in the next major but there's still quite a lot of work to do, and I have a bit less time right now with my newborn

But your idea will definitely land in the plugin, it's super nice to use!


> (this plugin with my auto center shenanigans has been so insanely robust, eye pleasing and most importantly pain free. I even get compliments about how clean my auto centering editor looks when I share my screen at work lol) 

❤️‍🔥❤️‍🔥 love to hear that, if you'd like any help to get started on contributing let me know

**Comment by Fabrice-Bernes** [comment]
Hi. This is a feature I'm interested in, but I wanted opinions before trying to make a pull request, because my solution back then was overcomplicated. Sadly I don't remember all the caveats I found back then, but here are some thoughts:

* We should have an option to center based on the longest line in the file, or the longest visible line.
* I don't know if re-centering every time a character is added to the buffer is too wasteful, so I was thinking about keeping track of the longest line found, and only resize if that limit is passed, or maybe just resize when leaving insert mode?. I don't know how to keep track of the longest line tho, and I remember having troubles with the opposite case (a line of the greatest lenght is shortened).
* I also remember having trouble with adding the lenght of the numbercolumn and the gutter
* Changing text without being in insert mode (e.g. with `J` in normal mode), resized NNP but fell short of a single character, so I simply added 5 characters to the result, but there's probably a way to get a proper number.

In any case,  here are some functions that resize using the longest line in the buffer, and the longest visible:
```lua
    local centerFull = function()
        local longest = 0
        local gutterwidth = vim.fn.getwininfo(vim.fn.win_getid())[1].textoff
        for _,line in pairs( vim.fn.getline(1, '$') ) do
            local current = vim.fn.strwidth( line )
            if current > longest then longest = current end
        end
        vim.cmd('NoNeckPainResize ' .. longest + gutterwidth) -- TODO add the length of signcolumn and linenr etc
        vim.g.NNP_length = longest + gutterwidth
    end

    local centerVisible = function()
        local longest = 0
        local gutterwidth = vim.fn.getwininfo(vim.fn.win_getid())[1].textoff
        for _,line in pairs( vim.fn.getline('w0', 'w$') ) do
            local current = vim.fn.strwidth( line )
            if current > longest then longest = current end
        end
        -- If we are not wraping, then add 1 to avoid the wrapsymbold
        -- And add 2 more for the cursor made when entering insert at the final character on a line
        vim.cmd('NoNeckPainResize ' .. longest + gutterwidth + 3) -- TODO add the length of signcolumn and linenr etc
        vim.g.NNP_length = longest
    end
```

Now, here are some autocommands that will help to keep the code centered as we type, scroll and change our buffer. I remember having trouble figuring out exactly what autocommand events we need to keep up with the many ways there are to change the text in a buffer, and how to know if there are more elements widening the buffer, like extmarks from lsp messages, or the foldcolumn, the gutter and so on.

```lua
    vim.api.nvim_create_autocmd({'WinScrolled'},{
        pattern = {'*'},
        callback = function() centerVisible() end
    })
    vim.api.nvim_create_autocmd({'TextChanged', 'InsertCharPre'},{
        callback = function()
            local current = vim.fn.strwidth(vim.fn.getline('.'))
            local gutterwidth = vim.fn.getwininfo(vim.fn.win_getid())[1].textoff
            if current+gutterwidth >= vim.g.NNP_length then
                -- Instead of adding 5 chars, one could try ctrl+o to run a normal command while in insert mode,
                -- and then ze to horizontally displace the text until the cursor is as far to the right at it can.
                -- Very hacky tho.
                vim.g.NNP_length = current + gutterwidth
                -- We add 5 to keep up with whatever is being added when the condition is met
                vim.cmd('NoNeckPainResize ' .. current + gutterwidth+5) -- TODO add the length of signcolumn and linenr etc
            end
        end
    })
    vim.api.nvim_create_autocmd({'InsertLeave'},{ callback = centerVisible })
```

Personally I prefer sticking to the neovim api, instead of using external programs like `wc`, but I don't know if it matters that much. Also I don't know how bad it is to do things this way instead of trying to use coroutines or stuff like that. Big screens and small fonts can fit a lot of lines, and there is a warning in the docs about being careful with the TextChanged autocommands, because of performance.

---------

As always, thank you for making and maintaining this plugin.
